### ISSUE TITLE
Weird behavior with losses

### ISSUE THREAD

**Issue by Hazarapet** [fchollet/keras#6680]
Guys I wanted to create a custom loss and noticed something weird.When I use my custom function (which returns the **same** keras' loss ) the accuracy is quite different.I think it's better to explain with code.
```
import keras
import numpy as np
import keras.backend as K
from keras import losses
from utils import components
from keras.models import Sequential
from keras.layers import Dense

def reg_binary_cross_entropy(y_true, y_pred):
    return losses.binary_crossentropy(y_true, y_pred)

model = Sequential()
model.add(Dense(4, activation='relu', input_dim=100))
model.add(Dense(3, activation='sigmoid'))

model.compile(optimizer='adam',
              loss=losses.binary_crossentropy,
              metrics=['accuracy'])
``
# Generate dummy data
data = np.random.random((10, 100)).astype(np.int8)
labels = np.zeros((10, 3)).astype(np.int8)

for i in range(labels.shape[0]):
    r_int = np.random.randint(3)
    labels[i][r_int] = 1

print 'training...'
# Train the model, iterating on the data in batches of 32 samples
loss = model.fit(data, labels, epochs=2, batch_size=32, verbose=1)
```

the `losses.binary_crossentropy or 'binary_crossentropy'` loss function returns `loss: 0.6931 - acc: 0.6667`, but when I use the function which returns the same calculation (reg_binary_cross_entropy), the accuracy starts to flap, here is what I've got `loss: 0.6931 - acc: 0.3000 or loss: 0.6931 - acc: 0.5000 or loss: 0.6931 - acc: 0.2000, loss: 0.6931 - acc: 0.1000`, which is so weird.I don't know, maybe the issue is not related with keras, but can you try this code and see what you got? and of course any idea about this issue.

**Comment by CA4GitHub** [comment]
@Hazarapet did you ever find a solution?

**Comment by Hazarapet** [comment]
No. I just avoided having that issue by changing the losses.

**Comment by sreenivasaupadhyaya** [comment]
Guys,

Did anyoe have a convincing solution to make custom_binarycrossentropy work?

I tried all possible methods (even making the whole training data size same as the bacth size to eliminate the dependence on global averaging during batch wise processing.). But i see significant difference between my binary cross entropy implementation and the one from keras ( by specifying loss = 'binary_crossentropy')

My crustom binary cross entropy code is as follows
def _loss_tensor(y_true, y_pred):
y_pred = K.clip(y_pred, _EPSILON, 1.0-_EPSILON)
out = (y_true * K.log(y_pred) + (1.0 - y_true) * K.log(1.0 - y_pred))
return -K.mean(out)
def _loss_tensor2(y_true, y_pred):
y_pred = K.clip(y_pred, _EPSILON, 1.0-_EPSILON)
out = -(y_true * K.log(y_pred) + -(1.0 - y_true) * K.log(1.0 - y_pred))
return out
def _loss_tensor2(y_true, y_pred):
loss1 = K.binary_crossentropy(y_true, y_pred)
return loss1
None of these methods work. It doesnt work even if i do K.mean() before ir eturn the results from custom loss function.

I am not able to understand what special does using loss = 'binary_crossentropy' does. When i use my custom loss function , the training sucks and it does work as expected.

I need my custom loss function to manipulate the loss function depending on the error and penalizing a certain type of classification error more.
